File: Algorithm/Python/400. Nth Digit.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-09-20 14:12:46
# @Last modified by:   WuLC
# @Last Modified time: 2016-09-20 14:12:57
# @Email: liangchaowu5@gmail.com


class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        count, base = 0, 1
        while (count + base * 9 * pow(10, base -1)) < n:
            count += base * 9 * pow(10, base -1)
            base += 1
        left = n - count
        num, mod = divmod(left, base)
        if mod:
            return int(str(pow(10,base-1) + num)[mod-1])
        else:
            return int(str(pow(10,base-1) + num -1)[mod-1])

File: Algorithm/Java/400. Nth Digit.java
/*
* @Author: WuLC
* @Date:   2017-09-10 08:57:52
* @Last Modified by:   WuLC
* @Last Modified time: 2017-09-10 15:03:48
*/


// be careful of overflow, use long instead of int

class Solution 
{
    public int findNthDigit(int n) 
    {
        int i = 1;
        while(true)
        {
            long tmp = (long)Math.pow(10, i-1) * 9 * i;
            if (n > tmp)
            {
                n -= tmp;
                i += 1;
            }
            else
            {
                int c = (n - 1) / i;
                int num = (int)Math.pow(10, i - 1) + c;
                return String.valueOf(num).charAt((n-1) % i) - '0';
            }
        }
        
    }
}

// more concise
class Solution 
{
	public int findNthDigit(int n) 
	{
		int len = 1;
		long count = 9;
		int start = 1;

		while (n > len * count) 
		{
			n -= len * count;
			len += 1;
			count *= 10;
			start *= 10;
		}

		start += (n - 1) / len;
		String s = Integer.toString(start);
		return Character.getNumericValue(s.charAt((n - 1) % len));
	}
}